array.go

```diff
@@ -1022,3 +1023,123 @@ func countElement(array []interface{}, element interface{}) int {
 	}
 	return count
 }
+
+// IsOrdered succeeds if array is ordered based on optional `comparator` function.
+// For default, it will use built-in comparator function for each data type.
+// It requires all elements in the array to have same data type
+//
+// Example:
+//
+//	array := NewArray(t, []interface{}{100, 101, 102})
+//	array.IsOrdered(func(x, y *Value) bool {
+//		valX, _ := x.Raw().(int)
+//		valY, _ := y.Raw().(int)
+//		return valX >= valY
+//	})
+func (a *Array) IsOrdered(less ...func(x, y *Value) bool) *Array {
+	a.chain.enter("IsOrdered()")
+	defer a.chain.leave()
+
+	if a.chain.failed() {
+		return a
+	}
+
+	if len(less) > 1 {
+		a.chain.fail(AssertionFailure{
+			Type: AssertUsage,
+			Errors: []error{
+				errors.New("unexpected multiple less arguments"),
+			},
+		})
+		return a
+	}
+
+	var fn func(x, y *Value) bool
+	if len(less) == 1 {
+		fn = less[0]
+	} else {
+		// use default comparator based on data type
+		cArr, ok := canonArray(a.chain, a.value)
+		if !ok {
+			return a
+		}
+
+		var currType, prevType string
+		for _, v := range cArr {
+			switch t := v.(type) {
+			case bool:
+				currType = fmt.Sprintf("%T", t)
+				fn = func(x, y *Value) bool {
+					// TODO: implement built-in comparator function for Boolean
+					return (!x.Boolean().value && y.Boolean().value) ||
+						(x.Boolean().value == y.Boolean().value)
+				}
+			case float64:
+				currType = fmt.Sprintf("%T", t)
+				fn = func(x, y *Value) bool {
+					x.Number().Le(y)
+					return !x.chain.failed()
+				}
+			case string:
+				currType = fmt.Sprintf("%T", t)
+				fn = func(x, y *Value) bool {
+					return x.String().value <= y.String().value
+				}
+			case time.Duration:
+				currType = fmt.Sprintf("%T", t)
+				fn = func(x, y *Value) bool {
+					valX, _ := x.value.(Duration)
+					valY, _ := y.value.(time.Duration)
+					valX.Le(valY)
+					return !x.chain.failed()
+				}
+			case time.Time:
+				currType = fmt.Sprintf("%T", t)
+				fn = func(x, y *Value) bool {
+					valX, _ := x.value.(DateTime)
+					valY, _ := y.value.(time.Time)
+					valX.Le(valY)
+					return !x.chain.failed()
+				}
```

Comment:
these 2 types cannot be type casted, since they're always detected as float64 and string respectively

Reply:
Could you elaborate a bit? Where they are being detected as float / string?

Reply:
Oh, that's because of canonArray, it converts them to floats.

Given it a second thought.. Array struct is a representation of JSON array, and that's why it uses canonArray. And in JSON, there are no Time and Duration. They can be encoded as strings or numbers, but we don't know about it and can't do any assumptions in default comparator.

Hence, it seems we just don't need to handle them by default. If needed, user can handle them in their compare function.